### PR TITLE
Add BM3D(sparse 3D transform-domain collaborative filtering) denoising algorithm.

### PR DESCRIPTION
**Pull request by Longhao-Chen**

### CODE REVIEW

**Comment by johnnychen94** [comment]
Thanks! Any reason to make it a conditional dependency?

Can you also add a demo similar to https://github.com/JuliaImages/ImageNoise.jl/blob/master/docs/examples/reduce_noise/nonlocal.jl?

Also, it would be great to also have a regression test to make sure future improvement don't make it worse, like how https://github.com/JuliaImages/ImageNoise.jl/blob/3381164f8a82cff6e2adc8cf1f10f83696b64755/test/ReduceNoise/NonlocalMean.jl#L62-L72 does.

FYI, usually, we use https://github.com/JuliaPackaging/Requires.jl for conditional codes

**Comment by Longhao-Chen** [comment]
Could you tell me what the role of conditional dependency is？

**Comment by johnnychen94** [comment]
- hard dependency: directly add `BM3DDenoise` to `Project.toml` and let Pkg to manage it
- conditional dependency: doesn't let Project.toml manage it, and only load it under some circumstance. In the current version, BM3DDenoise will only be loaded when `reduce_noise(img, BM3D())` is called.

Generally, we should use hard dependency and let Pkg do everything for us, because it allows 1) less code management, and 2) precompilation.

In some typical cases like `FileIO` there are too many possible IO backends, we use conditional loading to 1) don't run into a gigantic Project.toml because not everyone needs backend support for every file format, 2) reduce loading time

There're some discussions in https://github.com/JuliaLang/Pkg.jl/pull/977 to allow Pkg do conditional dependency management.

---

In our case, I think we can just use the hard dependency as if `BM3DNoise` is a normal package.

**Comment by Longhao-Chen** [comment]
I don't think everyone will use this function. For example, if I just want to use `apply_ noise`, then I don't have to install `BM3DDenoise.jl` , which can save disk space. Just like `FileIO` .
As for precompiling, I did an experiment. Precompiling will be performed during `BM3DDenoise.jl` installation.

**Comment by johnnychen94** [comment]
I realize that this package is a bit old and broken, so I've made a PR in #22 to upgrade it.

> I don't think everyone will use this function.

Okay, in this case, we can use Requires.jl for this purpose. We can write `BM3DDenoise.jl` in a usual way, and then conditionally load this file using Requires.

```julia
    # in src/ReduceNoise/ReduceNoise.jl
    using Requires

    function __init__()
        @require BM3DDenoise="95fb3b36-088a-43fb-bb1b-b1f34fadbd7d" include("BM3DDenoise.jl")
    end
```

**Comment by johnnychen94** [comment]
Oh sorry! Forget about the Requires stuff. I didn't get your ideas correctly. This is a nice idea and I think we should propose to other packages as well.

**Comment by johnnychen94** [comment]
Let's just add some basic demo and basic tests. And add BM3DDenoise to Project.toml

**Comment by Longhao-Chen** [comment]
> Let's just add some basic demo and basic tests. And add BM3DDenoise to Project.toml
OK, I will.

File: docs/examples/reduce_noise/BM3D.jl [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# title: The BM3D(sparse 3D transform-domain collaborative filtering) denoising algorithm.
+# id: bm3d-demo
+# cover: assets/bm3d_cover.png
+# date: 2021-08-04
+# author: Longhao Chen <Longhao.Chen@outlook.com>
+# description: This demo shows how to use the BM3D denoising algorithm to reduce gaussian noise
+# ---
```
Comment (johnnychen94) [inline docs/examples/reduce_noise/BM3D.jl:7]:
```suggestion
# ---
# title: The BM3D(sparse 3D transform-domain collaborative filtering) denoising algorithm.
# id: bm3d-demo
# cover: assets/bm3d_cover.png
# date: 2021-08-04
# author: "[Longhao Chen](https://github.com/Longhao-Chen)"
# description: Use the BM3D denoising algorithm to reduce gaussian noise
# ---
```

File: src/ReduceNoise/BM3DDenoise.jl [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+!!! info
+
+    This feature depends on the package: `BM3DDenoise`
+
+    You can install it by `Pkg`:
+    ```
+    using Pkg
+    Pkg.add("BM3DDenoise")
+    ```
```
Comment (johnnychen94) [inline src/ReduceNoise/BM3DDenoise.jl:25]: Actually, I was thinking of directly add `BM3DDenoise` into `Project.toml` but not load it until required by users. This way we still benefit from the lazy loading.

The source code of `BM3DDenoise` is quite small so preinstall it for all users sounds good to me.

How do you think?

Comment (Longhao-Chen) [reply src/ReduceNoise/BM3DDenoise.jl:25]: I have no problem with that.

Does the existing loading mechanism need to be modified?
Modify `Project.toml` only?

Comment (johnnychen94) [reply src/ReduceNoise/BM3DDenoise.jl:25]: Yep, I think we only need to modify `Project.toml` and remove this "!!! info" section.

File: Project.toml [hunk 3]
```diff
@@ -4,12 +4,15 @@ authors = ["Johnny Chen <johnnychen94@hotmail.com>"]
...
 [deps]
+BM3DDenoise = "95fb3b36-088a-43fb-bb1b-b1f34fadbd7d"
 Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
 ImageCore = "a09fc81d-aa75-5fe9-8630-4744c3626534"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
+UUIDs = "cf7118a7-6976-5b1a-9a39-7adc72f591a4"
 
 [compat]
+BM3DDenoise = "1.0"
```
Comment (johnnychen94) [inline Project.toml:15]:
```suggestion
BM3DDenoise = "1.0.1"
```

**Review by johnnychen94 — APPROVED** [review]

**Comment by johnnychen94** [comment]
@Longhao-Chen thank you for adding this!

**Pull request merged**